### PR TITLE
eth/catalyst: make getPayloadBodiesByRange take hex inputs

### PR DESCRIPTION
**Pull request by jwasinger**

### CODE REVIEW

**Comment by rjl493456442** [comment]
Spec says the parameters need to be `QUANTITY `, 

```
Values of a field of QUANTITY type MUST be encoded as a hexadecimal string with a 0x prefix and the leading 0s stripped (except for the case of encoding the value 0) matching the regular expression ^0x(?:0|(?:[a-fA-F1-9][a-fA-F0-9]*))$.

Note: Byte order of encoded value having QUANTITY type is big-endian.
```

Shouldn't we use `hexutil.Uint64()` instead?

**Comment by holiman** [comment]
We have , in `common.math`
```
// HexOrDecimal64 marshals uint64 as hex or decimal.
type HexOrDecimal64 uint64

``` 
as well as, in `rpc`
```
// DecimalOrHex unmarshals a non-negative decimal or hex parameter into a uint64.
type DecimalOrHex uint64
```

and `common.hexutil`
```
// Uint64 marshals/unmarshals as a JSON string with 0x prefix.
// The zero value marshals as "0x0".
type Uint64 uint64
```
Quite the assortment!

**Comment by fjl** [comment]
Removing `DecimalOrHex` in https://github.com/ethereum/go-ethereum/pull/26629

**Comment by fjl** [comment]
@jwasinger Please use `math.HexOrDecimal64` instead.

File: eth/catalyst/api.go [hunk 1]
```diff
@@ -772,20 +772,21 @@ func (api *ConsensusAPI) GetPayloadBodiesByHashV1(hashes []common.Hash) []*engin
...
 	if end > current {
 		end = current
 	}
-	var bodies []*engine.ExecutionPayloadBodyV1
-	for i := start; i < end; i++ {
+	var bodies []engine.ExecutionPayloadBodyV1
+	for i := uint64(start); i < end; i++ {
 		block := api.eth.BlockChain().GetBlockByNumber(i)
-		bodies = append(bodies, getBody(block))
+		body := getBody(block)
```
Comment (rjl493456442) [inline eth/catalyst/api.go:788]: Hmm, it's wrong. The returned body can be nil, isn't it?

File: eth/catalyst/api.go [hunk 2]
```diff
@@ -772,20 +772,21 @@ func (api *ConsensusAPI) GetPayloadBodiesByHashV1(hashes []common.Hash) []*engin
...
 	// limit count up until current
 	current := api.eth.BlockChain().CurrentBlock().NumberU64()
-	end := start + count
+	end := uint64(start) + uint64(count)
 	if end > current {
 		end = current
 	}
-	var bodies []*engine.ExecutionPayloadBodyV1
-	for i := start; i < end; i++ {
+	var bodies []engine.ExecutionPayloadBodyV1
```
Comment (MariusVanDerWijden) [inline eth/catalyst/api.go:785]:
```suggestion
	bodies := make([]*engine.ExecutionPayloadBodyV1, 0, uint64(count))
```
This way we should also pass the nil result vs empty result test

File: eth/catalyst/api.go (hunk 1 above)
Comment (holiman) [reply eth/catalyst/api.go:788]: > Hmm, it's wrong. The returned body can be nil, isn't it?

That's fine, we want a `nil` entry in the slice in that case (right?)

File: eth/catalyst/api.go [hunk 3]
```diff
@@ -772,20 +772,21 @@ func (api *ConsensusAPI) GetPayloadBodiesByHashV1(hashes []common.Hash) []*engin
...
 	current := api.eth.BlockChain().CurrentBlock().NumberU64()
-	end := start + count
+	end := uint64(start) + uint64(count)
 	if end > current {
 		end = current
 	}
-	var bodies []*engine.ExecutionPayloadBodyV1
-	for i := start; i < end; i++ {
+	bodies := make([]*engine.ExecutionPayloadBodyV1, 0, uint64(count))
+	for i := uint64(start); i < end; i++ {
```
Comment (holiman) [inline eth/catalyst/api.go:786]: A problem is here though. If we hit the `end > current` clause, then we need to make `i < end ` inclusive, to get `current` too. Or set `end = current +1`.

Comment (holiman) [reply eth/catalyst/api.go:786]: `start = 1, count = 2, current = 100`: should lead to `[1, 2]`
`start =1, count=100, current = 2`, should lead to `[1,2]`

File: eth/catalyst/api.go [hunk 4]
```diff
@@ -772,20 +772,21 @@ func (api *ConsensusAPI) GetPayloadBodiesByHashV1(hashes []common.Hash) []*engin
...
+func (api *ConsensusAPI) GetPayloadBodiesByRangeV1(start, count hexutil.Uint64) ([]engine.ExecutionPayloadBodyV1, error) {
 	if start == 0 || count == 0 || count > 1024 {
 		return nil, engine.InvalidParams.With(fmt.Errorf("invalid start or count, start: %v count: %v", start, count))
 	}
 	// limit count up until current
 	current := api.eth.BlockChain().CurrentBlock().NumberU64()
-	end := start + count
+	end := uint64(start) + uint64(count)
 	if end > current {
 		end = current
```
Comment (holiman) [inline eth/catalyst/api.go:783]:
```suggestion
		end = current + 1
```

**Review by holiman — APPROVED** [review]
LGTM

File: eth/catalyst/api.go (hunk 1 above)
Comment (MariusVanDerWijden) [reply eth/catalyst/api.go:788]: But the deref in line 789 should panic then, no?

**Comment by MariusVanDerWijden** [comment]
```diff
diff --git a/eth/catalyst/api.go b/eth/catalyst/api.go
index cf67e19d5..15b53985a 100644
--- a/eth/catalyst/api.go
+++ b/eth/catalyst/api.go
@@ -772,7 +772,7 @@ func (api *ConsensusAPI) GetPayloadBodiesByHashV1(hashes []common.Hash) []*engin
 
 // GetPayloadBodiesByRangeV1 implements engine_getPayloadBodiesByRangeV1 which allows for retrieval of a range
 // of block bodies by the engine api.
-func (api *ConsensusAPI) GetPayloadBodiesByRangeV1(start, count hexutil.Uint64) ([]engine.ExecutionPayloadBodyV1, error) {
+func (api *ConsensusAPI) GetPayloadBodiesByRangeV1(start, count hexutil.Uint64) ([]*engine.ExecutionPayloadBodyV1, error) {
        if start == 0 || count == 0 || count > 1024 {
                return nil, engine.InvalidParams.With(fmt.Errorf("invalid start or count, start: %v count: %v", start, count))
        }
@@ -782,11 +782,10 @@ func (api *ConsensusAPI) GetPayloadBodiesByRangeV1(start, count hexutil.Uint64)
        if last > current {
                last = current
        }
-       bodies := make([]engine.ExecutionPayloadBodyV1, 0, uint64(count))
+       bodies := make([]*engine.ExecutionPayloadBodyV1, 0, uint64(count))
        for i := uint64(start); i <= last; i++ {
                block := api.eth.BlockChain().GetBlockByNumber(i)
-               body := getBody(block)
-               bodies = append(bodies, *body)
+               bodies = append(bodies, getBody(block))
        }
        return bodies, nil
 }
diff --git a/eth/catalyst/api_test.go b/eth/catalyst/api_test.go
index 7b01d5843..cbd309a01 100644
--- a/eth/catalyst/api_test.go
+++ b/eth/catalyst/api_test.go
@@ -1367,7 +1367,7 @@ func TestGetBlockBodiesByRange(t *testing.T) {
                }
                if len(result) == len(test.results) {
                        for i, r := range result {
-                               if !equalBody(test.results[i], &r) {
+                               if !equalBody(test.results[i], r) {
                                        t.Fatalf("test %d: invalid response: expected \n%+v\ngot\n%+v", k, test.results[i], r)
                                }
                        }
```

**Comment by holiman** [comment]
@MariusVanDerWijden yes please

File: eth/catalyst/api.go (hunk 1 above)
Comment (holiman) [reply eth/catalyst/api.go:788]: >  I don't understand why we need to make the result an array not an array of pointers

I agree

**Review by MariusVanDerWijden — APPROVED** [review]
LGTM

File: eth/catalyst/api_test.go [hunk 5]
```diff
@@ -1332,31 +1337,42 @@ func TestGetBlockBodiesByRange(t *testing.T) {
...
+		// whole range unavailable
+		{
+			results: make([]*types.Body, 0),
+			start:   22,
 			count:   2,
 		},
 	}
 
-	for k, test := range tests {
+	for k, test := range tests[:] {
```
Comment (holiman) [inline eth/catalyst/api_test.go:1363]:
```suggestion
	for k, test := range tests {
```

File: eth/catalyst/api.go (hunk 1 above)
Comment (jwasinger) [reply eth/catalyst/api.go]: it was a shot-in-the-dark attempt at a fix.

**Pull request merged**